posts/opaleyes-api-breakage-policy.md

```diff
@@ -0,0 +1,385 @@
+I used the same technique above of defining a type synonym
+
+```haskell
+type DefaultFromField = QueryRunnerColumnDefault
+```
+
+The synonym could be used in most places that original class name
+could be used, but *not* in instance heads.  For example, I can write
+the following, but I cannot replace `QueryRunnerColumnDefault` in the
+instance head with `DefaultFromField`.
+
+```haskell
+instance DefaultFromField a b
+   => QueryRunnerColumnDefault (Nullable a) (Maybe b) where
+   ...
+```
+
+Perhaps this is a weakness in the language.  It certainly prevents a
+smooth deprecation cycle for class names.
+
+## Weaknesses
+
+It's not always possible to avoid flag day and it's not always
+possible to uphold the "fundamental principle".  For example, the type
+class synonym example above shows that it is impossible to avoid a
+flag day when renaming classes.  In version `N` users would *have to*
+use `QueryRunnerColumnDefault` in type class instance heads, and in
+version `N+1` they would *have to* use `DefaultFromField`.
+Regrettably there is no way to write code that works across version
+`N` and `N+1`[^cpp].  Perhaps a change to GHC could add support for
+this workflow.
+
+Another example is a change to Opaleye's type of SQL fields or
+columns, that used to be called `Column`.  Non-nullable columns were
+represented as, for example, `Column SqlText` and nullable columns
+were represented as `Column (Nullable Sql)`.  Unfortunately the
+nonsensical `Column (Nullable (Nullable SqlText))` was not prevented.
+To tighten up the types, version 0.9 improved the situation by
+replacing `Column` with `Field_`[^field], which has the distinction
+`Field_ NonNullable SqlText` vs. `Field_ Nullable SqlText`.  Replacing
+`Column` with `Field_` caused a flag day for the Opaleye ecosystem.
+Although there is [a compatibility type
+family](https://hackage.haskell.org/package/opaleye-0.9.2.0/docs/Opaleye-Column.html#t:Column)
+it's not enough to guarantee that user code can work across 0.8 and
+0.9 at once.
+
+## Summary
+
+There is a wide spectrum between "never break userspace" and "move
+fast and break things"[^slogans] and the right place on that spectrum
+for a project to position itself can only be determined on a
+case-by-case basis by taking into account the needs of the project
+maintainers and users.  If the project decides that some rate of API
+breakage is desirable then the maintainers can make life easier for
+the users by following a principle like Opaleye's fundamental
+principle if API breakage.  Users' lives are made easier if they are
+given breadcrumbs and hand-holding when their code breaks.
+
+Do you have thoughts or comments on this article?  Do you know
+articles explaining similar API breakage policies?  Or different
+policies?  Would you like my help designing a deprecation cycle for
+your library?  Feel free to [contact
+me](http://web.jaguarpaw.co.uk/~tom/contact/).
+
+
```

Comment:
(randomly placed thread)

Since opaleye has haskell-external dependencies: If opaleye would benefit from features only available in a new version of postgres, how would that be dealt with?

Reply:
This does occasionally happen but it tends to be limited to individual functions which the user can use or not at their preference, depending on the version of Postgres they use.  Even pervasive things like `LATERAL` can be chosen by using monad notation vs arrow notation.